git_machete/git_operations.py

```diff
@@ -206,22 +208,33 @@ def __init__(self) -> None:
 
     @staticmethod
     def _run_git(git_cmd: str, *args: str, **kwargs: Any) -> int:
-        exit_code = utils.run_cmd("git", git_cmd, *args, **kwargs)
-        if not kwargs.get("allow_non_zero") and exit_code != 0:
-            raise MacheteException(f"`{utils.get_cmd_shell_repr('git', git_cmd, *args, env=kwargs.get('env'))}` returned {exit_code}")
+        kwargs_ = kwargs.copy()
+        allow_non_zero = kwargs_.pop("allow_non_zero", False)
+        exit_code = utils.run_cmd("git", git_cmd, *args, **kwargs_)
+        if not allow_non_zero and exit_code != 0:
+            raise MacheteException(f"`{utils.get_cmd_shell_repr('git', git_cmd, *args, env=kwargs_.get('env'))}` returned {exit_code}")
         return exit_code
 
     @staticmethod
-    def _popen_git(git_cmd: str, *args: str, **kwargs: Any) -> str:
-        allow_non_zero = kwargs.pop("allow_non_zero", False)
-        exit_code, stdout, stderr = utils.popen_cmd("git", git_cmd, *args, **kwargs)
+    def _popen_git(git_cmd: str, *args: str, **kwargs: Any) -> Tuple[int, str, str]:
+        kwargs_ = kwargs.copy()
+        allow_non_zero = kwargs_.pop("allow_non_zero", False)
+        exit_code, stdout, stderr = utils.popen_cmd("git", git_cmd, *args, **kwargs_)
         if not allow_non_zero and exit_code != 0:
-            exit_code_msg: str = fmt(f"`{utils.get_cmd_shell_repr('git', git_cmd, *args, env=kwargs.get('env'))}` returned {exit_code}\n")
+            exit_code_msg: str = fmt(f"`{utils.get_cmd_shell_repr('git', git_cmd, *args, env=kwargs_.get('env'))}` returned {exit_code}\n")
             stdout_msg: str = f"\n{utils.bold('stdout')}:\n{utils.dim(stdout)}" if stdout else ""
             stderr_msg: str = f"\n{utils.bold('stderr')}:\n{utils.dim(stderr)}" if stderr else ""
             # Not applying the formatter to avoid transforming whatever characters might be in the output of the command.
             raise MacheteException(exit_code_msg + stdout_msg + stderr_msg, apply_fmt=False)
-        return stdout
+        return exit_code, stdout, stderr
+
+    @staticmethod
+    def _popen_git_stdout(git_cmd: str, *args: str, **kwargs: Any) -> str:
```

Comment:
Hmmm the problem is that "popen" itself indicates that stdout is returned/available 🤔 so I'd rename further the methods, maybe sth like:
`_run_git` -> `_run_git_output_to_console`
`_popen_git_stdout` -> `_run_git_returning_stdout`
`_popen_git_exit_code` -> `_run_git_discarding_output`
❓

Reply:
Or indeed make `_popen_git` returning a named tuple with `stdout`, `stderr`, `exit_code`? then it's gonna be super-clear & consistent

Reply:
> Or indeed make `_popen_git` return a named tuple with `stdout`, `stderr`, `exit_code`? then it's gonna be super-clear & consistent

@PawelLipski Hmm, okay, lets go with it